Sources/CoreMetrics/Metrics.swift

```diff
@@ -12,175 +12,250 @@
+        self.recordNanoseconds(Int64(duration) * 1_000_000_000)
     }
 
     @inlinable
-    func withGauge(label: String, dimensions: [(String, String)] = [], then: (Recorder) -> Void) {
-        then(self.makeGauge(label: label, dimensions: dimensions))
+    public func recordSeconds<DataType: BinaryFloatingPoint>(_ duration: DataType) {
+        self.recordNanoseconds(Int64(duration * 1_000_000_000))
     }
 }
 
-public enum Metrics {
-    private static let lock = ReadWriteLock()
-    private static var _handler: MetricsHandler = NOOPMetricsHandler.instance
-
-    public static func bootstrap(_ handler: MetricsHandler) {
-        self.lock.withWriterLockVoid {
-            self._handler = handler
-        }
+public extension Timer {
+    public convenience init(label: String, dimensions: [(String, String)] = []) {
+        let handler = MetricsSystem.factory.makeTimer(label: label, dimensions: dimensions)
+        self.init(label: label, dimensions: dimensions, handler: handler)
     }
+}
 
-    public static var global: MetricsHandler {
-        return self.lock.withReaderLock { self._handler }
+public extension Timer {
+    @inlinable
+    public static func `do`(label: String, dimensions: [(String, String)] = [], body: (Timer) -> Void) {
+        body(Timer(label: label, dimensions: dimensions))
     }
 }
 
-public extension Metrics {
-    @inlinable
-    func makeCounter(label: String, dimensions: [(String, String)]) -> Counter {
-        return Metrics.global.makeCounter(label: label, dimensions: dimensions)
+public protocol MetricsFactory {
+    func makeCounter(label: String, dimensions: [(String, String)]) -> CounterHandler
+    func makeRecorder(label: String, dimensions: [(String, String)], aggregate: Bool) -> RecorderHandler
+    func makeTimer(label: String, dimensions: [(String, String)]) -> TimerHandler
+}
+
+// This is the metrics system itself, it's mostly used set the type of the `MetricsFactory` implementation.
+public enum MetricsSystem {
+    fileprivate static let lock = ReadWriteLock()
+    fileprivate static var _factory: MetricsFactory = NOOPMetricsHandler.instance
+    fileprivate static var initialized = false
+
+    // Configures which `LogHandler` to use in the application.
+    public static func bootstrap(_ factory: MetricsFactory) {
+        self.lock.withWriterLock {
+            precondition(!self.initialized, "metrics system can only be initialized once per process. currently used factory: \(self.factory)")
+            self._factory = factory
+            self.initialized = true
+        }
     }
 
-    @inlinable
-    func makeRecorder(label: String, dimensions: [(String, String)], aggregate: Bool) -> Recorder {
-        return Metrics.global.makeRecorder(label: label, dimensions: dimensions, aggregate: aggregate)
+    // for our testing we want to allow multiple bootstraping
+    internal static func bootstrapInternal(_ factory: MetricsFactory) {
```

Comment:
Since this should only ever be used for testing purposes, and not really internally, would it not make more sense to call this `bootstrapTestable` or something of the likes?

Reply:
@MrLotU fwiw we have the same thing in the logging API, so whatever we choose we should stay consistent. i'd suggest we merge this and then opening a focused PR to start such conversation. wdyt?

Reply:
Ah, in that case choice of name makes more sense. Keeping it like this and opening up a discussion later sounds good 👍